procedures/unit-testing-utils.ipf

```diff
@@ -893,68 +908,167 @@ static Function NumBytesInUTF8Character(str, byteOffset)
+			result.v2 = prefix + GetStringWithContext(str2, start, start + diffpos, end2 - 1)
+			return NaN
+		endif
+
+		// one line is shorter than the other
+		if(end1 != end2)
+			sprintf prefix, "%d:%d:%d>", line, endmin - start, endmin
+			result.v1 = prefix + GetStringWithContext(str1, start, endmin, end1)
+			result.v2 = prefix + GetStringWithContext(str2, start, endmin, end2)
+			return NaN
+		endif
+
+		// the line endings are different
+		if(!!CmpStr(marker1, marker2))
+			sprintf prefix, "%d:%d:%d>", line, endmin - start, endmin
+			result.v1 = prefix + GetStringWithContext(str1, start, endmin, endmin + strlen(marker1) - 1)
+			result.v2 = prefix + GetStringWithContext(str2, start, endmin, endmin + strlen(marker2) - 1)
+			return NaN
+		endif
+
+		// no differences in this line
+		start = endmin + strlen(marker1)
+		line += 1
+
+	while(start<strlen(str1) && start<strlen(str2))
+
+	// one string is larger than the other one
+	if(strlen(str1) != strlen(str2))
+		sprintf prefix, "%d:0:%d>", line, start
+		if(strlen(str1) <= start)
+			result.v1 = prefix
+		else
+			end1 = DetectEndOfLine(str1, start, marker1)
+			result.v1 = prefix + EscapeString(str1[start, end1])
+		endif
+		if(strlen(str2) <= start)
+			result.v2 = prefix
+		else
+			end2 = DetectEndOfLine(str2, start, marker2)
+			result.v2 = prefix + EscapeString(str2[start, end2])
+		endif
+		return NaN
+	endif
+
+	Abort("Cannot create diff of equal strings")
+End
+
+static Function/S GetStringWithContext(str, start, diffpos, endpos)
+	variable start, diffpos, endpos
 	string str
 
-	variable bytePos, charNum, numBytes, code
-	variable extPrint, len
-	string format, hexOut
-	string extOut = ""
-
-	Make/FREE/I/U biDiChars = { \
-							char2num("\u200E"), char2num("\u200F"), char2num("\u061C"), char2num("\u202A"), \
-							char2num("\u202D"), char2num("\u202B"), char2num("\u202E"), char2num("\u202C"), \
-							char2num("\u2066"), char2num("\u2067"), char2num("\u2068"), char2num("\u2069")}
-
-	len = strlen(str)
-	if(len)
-		do
-			numBytes = NumBytesInUTF8Character(str, bytePos)
-			code = char2num(str[bytePos, bytePos + numBytes - 1])
-			if(code >= char2num("\u0000") && code <= char2num("\u001F"))
-				// C0 controls
-				extPrint = 1
-			elseif(code >= char2num("\u007F") && code <= char2num("\u009F"))
-				// C1 controls
-				extPrint = 1
+	string strOut
+
+	strOut = EscapeString(str[max(start, diffpos - MAX_STRING_DIFF_CONTEXT), diffpos - 1])
+	strOut += EscapeString(str[diffpos, min(endpos, diffpos + MAX_STRING_DIFF_CONTEXT)])
+	return strOut
+End
+
+static Function GetTextDiffPos(str1, str2)
+	string str1, str2
+
+	variable i
+
+	for(i=0; i < strlen(str1); i += 1)
+		if (!!CmpStr(str1[i], str2[i]))
+			return i
+		endif
+	endfor
+
+	return -1
+End
+
+static Function DetectEndOfLine(str, start, marker)
+	variable start
+	string str, &marker
+
+	variable i
+
+	for(i = start; i < strlen(str); i += 1)
+		if (!CmpStr(str[i], "\r"))
+			if (!CmpStr(str[i + 1], "\n"))
```

Comment:
This was an intended use of the Igor Feature. In other programming languages i would check the length of the string first to not read beyond its length.

I can add this check but its kind of duplicate because of the Igor behavior.